Modules/Loadable/Markups/Logic/vtkSlicerMarkupsLogic.cxx

```diff
@@ -1561,3 +1561,63 @@ vtkMRMLMarkupsJsonStorageNode* vtkSlicerMarkupsLogic::AddNewJsonStorageNodeForMa
 {
   return vtkMRMLMarkupsJsonStorageNode::SafeDownCast(this->GetMRMLScene()->AddNewNodeByClass(this->GetJsonStorageNodeClassNameForMarkupsType(markupsType)));
 }
+
+//---------------------------------------------------------------------------
+void vtkSlicerMarkupsLogic::SetMarkup(vtkMRMLMarkupsNode *markupsNode,
+                                      vtkSlicerMarkupsWidget* markupsWidget)
+{
+  if (markupsNode == nullptr)
+    {
+    vtkErrorMacro("RegisterMarkup: Invalid node.");
+    return;
+    }
+
+  if (markupsWidget == nullptr)
+    {
+    vtkErrorMacro("RegisterMarkup: Invalid widget.");
+    return;
+    }
+
+  // Take ownership of widget regardless of the outcome of this function (we will
+  // manage the memory).
+  vtkSmartPointer<vtkMRMLMarkupsNode> node =
+    vtkSmartPointer<vtkMRMLMarkupsNode>::Take(markupsNode);
+
+  vtkSmartPointer<vtkSlicerMarkupsWidget> associatedWidget =
+    vtkSmartPointer<vtkSlicerMarkupsWidget>::Take(markupsWidget);
+
+  // Check that the class is not already registered
+  if (this->MarkupsWidgetsMap.count(markupsNode->GetMarkupName()))
+    {
+    vtkWarningMacro("RegisterMarkup: Markups node " << markupsNode->GetMarkupName() <<
+                    " is already registered.");
+    return;
+    }
+
+  this->MarkupsWidgetsMap[markupsNode->GetMarkupName()] = associatedWidget;
```

Comment:
Name of map variable should contain both what maps to what, something like MarkupsTypeToWidgetMap and MarkupsTypeToNodeMap. It would probably also better to map the markups type to a struct because you always need to add/remove widget and node together. It would be also easier to specify additional properties in the future (you would not need to create a new map for each additional property).